### PR TITLE
grpc: bi-directional streaming region heartbeats 

### PR DESCRIPTION
**Pull request by overvenus**
This PR replaces unary region heartbeats with bi-directional streaming heartbeats, it intends to increase heartbeats throughput and enable server push in the future.

Pending pingcap/kvproto#152

### CODE REVIEW

File: server/grpc_service.go [hunk 1]
```diff
@@ -234,66 +234,74 @@ func (s *Server) StoreHeartbeat(ctx context.Context, request *pdpb.StoreHeartbea
...
 
-	resp := &pdpb.RegionHeartbeatResponse{}
-	region := newRegionInfo(request.GetRegion(), request.GetLeader())
-	region.DownPeers = request.GetDownPeers()
-	region.PendingPeers = request.GetPendingPeers()
-	if region.GetId() == 0 {
-		pberr := &pdpb.Error{
-			Type:    pdpb.ErrorType_UNKNOWN,
-			Message: fmt.Sprintf("invalid request region, %v", request),
+		var region *RegionInfo
```
Comment (siddontang) [inline server/grpc_service.go]: don't use `goto`
extract a function which to handle the heartbeat

File: server/grpc_service.go [hunk 2]
```diff
@@ -234,66 +234,90 @@ func (s *Server) StoreHeartbeat(ctx context.Context, request *pdpb.StoreHeartbea
...
 
-	resp := &pdpb.RegionHeartbeatResponse{}
-	region := newRegionInfo(request.GetRegion(), request.GetLeader())
-	region.DownPeers = request.GetDownPeers()
-	region.PendingPeers = request.GetPendingPeers()
-	if region.GetId() == 0 {
-		pberr := &pdpb.Error{
-			Type:    pdpb.ErrorType_UNKNOWN,
-			Message: fmt.Sprintf("invalid request region, %v", request),
+		resp := &pdpb.RegionHeartbeatResponse{}
```
Comment (siddontang) [inline server/grpc_service.go]: As I said before, please extract a function which returns `(pdpb.RegionHeartbeatResponse, error)`, then send the response.

**Comment by siddontang** [comment]
any update?

File: server/grpc_service.go [hunk 3]
```diff
@@ -234,67 +234,94 @@ func (s *Server) StoreHeartbeat(ctx context.Context, request *pdpb.StoreHeartbea
...
-	if region.Leader == nil {
-		pberr := &pdpb.Error{
-			Type:    pdpb.ErrorType_UNKNOWN,
-			Message: fmt.Sprintf("invalid request leader, %v", request),
+		region := newRegionInfo(request.GetRegion(), request.GetLeader())
+		region.DownPeers = request.GetDownPeers()
+		region.PendingPeers = request.GetPendingPeers()
+		region.WrittenBytes = request.GetBytesWritten()
+		if region.GetId() == 0 {
+			pberr := &pdpb.Error{
```
Comment (siddontang) [inline server/grpc_service.go]: duplicated Unknown error handling, can you extract these to a function?

**Comment by siddontang** [comment]
PTAL @disksing

**Comment by siddontang** [comment]
CI failed @overvenus

File: server/api/history_test.go [hunk 4]
```diff
@@ -14,11 +14,13 @@
 package api
 
 import (
+	"context"
```
Comment (disksing) [inline server/api/history_test.go]: Use "golang.org/x/net/context" instead.

File: server/grpc_service.go [hunk 5]
```diff
@@ -234,67 +234,88 @@ func (s *Server) StoreHeartbeat(ctx context.Context, request *pdpb.StoreHeartbea
...
+		}
+		if err != nil {
+			return errors.Trace(err)
+		}
 
-	cluster := s.GetRaftCluster()
-	if cluster == nil {
-		return &pdpb.RegionHeartbeatResponse{Header: s.notBootstrappedHeader()}, nil
-	}
+		// TODO: should we check headers here?
```
Comment (disksing) [inline server/grpc_service.go]: Yes, we should.

**Comment by disksing** [comment]
LGTM.

File: server/config.go [hunk 6]
```diff
@@ -80,17 +80,19 @@ type Config struct {
...
 
 	configFile string
 
 	// For all warnings during parsing.
 	WarningMsgs []string
+
+	// Only test can change them.
+	nextRetryDelay             time.Duration
+	disableStrictReconfigCheck bool
+	// It is a workaround for passing tests,which make test region heartbeat easy.
```
Comment (siddontang) [inline server/config.go]: which makes

**Comment by siddontang** [comment]
PTAL @huachaohuang

**Comment by huachaohuang** [comment]
Not quite sure about the grpc things for now, others looks good to me.

**Comment by huachaohuang** [comment]
CI failed.

File: server/grpc_service.go [hunk 7]
```diff
@@ -234,67 +234,91 @@ func (s *Server) StoreHeartbeat(ctx context.Context, request *pdpb.StoreHeartbea
...
+			msg := errors.Trace(err).Error()
+			err = sendErrorRegionHeartbeatResponse(server, s.clusterID, pdpb.ErrorType_UNKNOWN, msg)
+			if err != nil {
+				return errors.Trace(err)
+			}
+			continue
+		}
+		if resp == nil {
+			if s.cfg.regionHeartbeatUnaryMode {
+				// A workaround for passing tests, remove it ASAP.
```
Comment (siddontang) [inline server/grpc_service.go]: do we still need it?

Comment (overvenus) [inline server/grpc_service.go]: Yes, for passing tests, current tests still stay in unary model, one request one response, I will refactor those tests later.

**Pull request merged**